source/library/Argo/Codec.hs

```diff
@@ -0,0 +1,284 @@
+    (State.State Value.Value)
+    a
+
+valueCodec :: ValueCodec Value.Value
+valueCodec = Codec
+    { decode = Reader.ask
+    , encode = \ x -> do
+        State.put x
+        pure x
+    }
+
+nullCodec :: ValueCodec Null.Null
+nullCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Null y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Null but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Null x
+        pure x
+    }
+
+booleanCodec :: ValueCodec Boolean.Boolean
+booleanCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Boolean y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Boolean but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Boolean x
+        pure x
+    }
+
+numberCodec :: ValueCodec Number.Number
+numberCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Number y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Number but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Number x
+        pure x
+    }
+
+stringCodec :: ValueCodec String.String
+stringCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.String y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected String but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.String x
+        pure x
+    }
+
+arrayCodec :: ValueCodec (Array.ArrayOf Value.Value)
+arrayCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Array y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Array but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Array x
+        pure x
+    }
+
+objectCodec :: ValueCodec (Object.ObjectOf Value.Value)
+objectCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Object y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Object but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Object x
+        pure x
+    }
+
+boolCodec :: ValueCodec Bool
+boolCodec = dimap Boolean.toBool Boolean.fromBool booleanCodec
+
+textCodec :: ValueCodec Text.Text
+textCodec = dimap String.toText String.fromText stringCodec
+
+eitherCodec :: ValueCodec a -> ValueCodec b -> ValueCodec (Either a b)
+eitherCodec c1 c2 = Codec
+    { decode = fmap Left (decode c1) <|> fmap Right (decode c2)
+    , encode = \ x -> case x of
+        Left y -> fmap Left $ encode c1 y
+        Right y -> fmap Right $ encode c2 y
+    }
+
+maybeCodec :: ValueCodec a -> ValueCodec (Maybe a)
+maybeCodec =
+    dimap
+        (either Just $ const Nothing)
+        (maybe (Right $ Null.fromUnit ()) Left)
+    . flip eitherCodec nullCodec
+
+data Permission
+    = Allow
+    | Forbid
+    deriving (Eq, Show)
```

Comment:
I'm pretty pleased with this data type for handling excess elements in arrays and excess members in objects. It feels better to me than Aeson's [`rejectUnknownFields`](https://hackage.haskell.org/package/aeson-2.0.1.0/docs/Data-Aeson.html#v:rejectUnknownFields).